docs_src/advanced/background_tasks.py

```diff
@@ -0,0 +1,57 @@
+from typing import AsyncIterator, List
+
+import anyio
+import anyio.abc
+
+from xpresso import App, Depends, Path
+from xpresso.typing import Annotated
+
+
+async def create_conn_task_group() -> AsyncIterator[
+    anyio.abc.TaskGroup
+]:
+    async with anyio.create_task_group() as tg:
+        try:
+            yield tg
+        except Exception as e:
+            print(e)
+            pass
+        print("exiting")
+
+
+ConnectionTaskGroup = Annotated[
+    anyio.abc.TaskGroup,
+    Depends(create_conn_task_group, scope="connection"),
+]
+
+
+async def create_app_task_group() -> AsyncIterator[
+    anyio.abc.TaskGroup
+]:
+    async with anyio.create_task_group() as tg:
+        try:
+            yield tg
+        except Exception as e:
+            print(e)
+            pass
+        print("exiting")
+
+
+AppTaskGroup = Annotated[
+    anyio.abc.TaskGroup, Depends(create_app_task_group, scope="app")
+]
```

Comment:
This exposes two possible deficiencies of `di`:

## You can't use the same dependency (callable) in two different scopes within the same dependency graph.

So we can't do this

```python
async def create_task_group() -> AsyncIterator[
     anyio.abc.TaskGroup
 ]:
     async with anyio.create_task_group() as tg:
         try:
             yield tg
         except Exception as e:
             print(e)
             pass
         print("exiting")


 ConnectionTaskGroup = Annotated[
     anyio.abc.TaskGroup,
     Depends(create_task_group, scope="connection"),
 ]

 AppTaskGroup = Annotated[
     anyio.abc.TaskGroup, Depends(create_task_group, scope="app")
  ]
```

This is tricky to solve because in some cases you _want_ to treat dependencies in different scopes as the same dependency, and in others you don't.
I guess we could change this back, it would mean this example:

```python
async def lifespan(log: Log):
    ....

async def endpoint(log: Log):
    ....
```

Becomes:

```python
async def lifespan(log: Log):
    ....

async def endpoint(log: Annotated[Log, Depends(scope="app")]):
    ....
```

## context manager dependencies have to a be a generator (not a function returning a context manager)

So we can't do this

```python
 ConnectionTaskGroup = Annotated[
     anyio.abc.TaskGroup,
     Depends(anyio.create_task_group, scope="connection"),
 ]

 AppTaskGroup = Annotated[
     anyio.abc.TaskGroup, Depends(anyio.create_task_group, scope="app")
  ]
```

This relates to introspection in `di`. For performance (code specialization) and ease of implementing an executor, `di` determines the kind of dependency _before_ it ever calls it.
So we can't introspect into a callable returning a context manager (that's what `anyio.create_task_group` does).
We could defer this introspection to runtime (call the thing and check what it returns), but this presents 2 issue:
1. Complexity. If we want to keep the performance of code specialization, we need a first pass that handles all cases and then replaces itself with a specialized version. We'd also need to change the executor API so that it has no option but to call `inspect.isawaitable` on every dependency instead of checking an attribute (as it currently does).
2. Reliability. What happens if the callable returns a context manager the first time and something else the second? That's a horrible design but I guess it's possible. There's no way we want to introspect every time, so we'd have to prohibit this (seems reasonable). But this may be a red-herring of unsoundness of the introspect after calling approach.